src/sockaddr.rs

```diff
@@ -310,40 +310,192 @@ impl fmt::Debug for SockAddr {
     }
 }
 
-#[test]
-fn ipv4() {
-    use std::net::Ipv4Addr;
-    let std = SocketAddrV4::new(Ipv4Addr::new(1, 2, 3, 4), 9876);
-    let addr = SockAddr::from(std);
-    assert_eq!(addr.family(), AF_INET as sa_family_t);
-    assert_eq!(addr.len(), size_of::<sockaddr_in>() as socklen_t);
-    assert_eq!(addr.as_socket(), Some(SocketAddr::V4(std)));
-    assert_eq!(addr.as_socket_ipv4(), Some(std));
-    assert!(addr.as_socket_ipv6().is_none());
-
-    let addr = SockAddr::from(SocketAddr::from(std));
-    assert_eq!(addr.family(), AF_INET as sa_family_t);
-    assert_eq!(addr.len(), size_of::<sockaddr_in>() as socklen_t);
-    assert_eq!(addr.as_socket(), Some(SocketAddr::V4(std)));
-    assert_eq!(addr.as_socket_ipv4(), Some(std));
-    assert!(addr.as_socket_ipv6().is_none());
+impl cmp::PartialEq for SockAddr {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        let this_sockaddr = self.as_ptr();
+        let other_sockaddr = other.as_ptr();
+        unsafe {
+            (*this_sockaddr).sa_family == (*other_sockaddr).sa_family
+                && (*this_sockaddr).sa_data == (*other_sockaddr).sa_data
```

Comment:
This is UB for short address. Please look at the `as_socket` method and only access the initialised data.

Reply:
@Thomasdezeeuw 
What do I do in case `ss_family` is neither `AF_INET` nor `AF_INET6`... do I panic?
```
/// Returns this address as a `SocketAddr` if it is in the `AF_INET` (IPv4)
    /// or `AF_INET6` (IPv6) family, otherwise returns `None`.
    pub fn as_socket(&self) -> Option<SocketAddr> {
        if self.storage.ss_family == AF_INET as sa_family_t {
...
        } else if self.storage.ss_family == AF_INET6 as sa_family_t {
...
        } else {
            None
        }
    }
```

```
pub struct SockAddr {
    storage: sockaddr_storage,
    len: socklen_t,
}
```
we know the `len`gth of the `storage: sockaddr_storage`, can I simply compare raw bytes?